aws/acm.go

```diff
@@ -1,227 +1,69 @@
 			certList = append(certList, cert)
 		}
-		return true // never stop iterator
+		return true
 	})
-
 	return certList, err
 }
 
-// FindBestMatchingCertificate get all ACM certificates and use a suffix
-// search best match operation in order to find the best matching
-// certificate ARN.
-//
-// We don't need to validate the Revocation here, because we only pull
-// ISSUED certificates.
-func FindBestMatchingCertificate(certs []*CertDetail, hostname string) (*CertDetail, error) {
-	candidate := &CertDetail{}
-	longestMatch := -1
-	now := time.Now()
-
-	for _, cert := range certs {
-		notAfter := cert.NotAfter
-		notBefore := cert.NotBefore
-
-		// ignore invalid timeframes
-		if now.After(notAfter) || now.Before(notBefore) {
-			continue
-		}
-
-		for _, altName := range cert.AltNames {
-			if prefixGlob(altName, hostname) {
-				nameLength := len(altName)
-
-				switch {
-				case longestMatch < 0:
-					// first matching found
-					longestMatch = nameLength
-					candidate = cert
-				case longestMatch < nameLength:
-					if notBefore.Before(now) && notAfter.Add(-minimalCertValidityPeriod).After(now) {
-						// more specific valid cert found: *.example.org -> foo.example.org
-						longestMatch = nameLength
-						candidate = cert
-					}
-				case longestMatch == nameLength:
-					if notBefore.After(candidate.NotBefore) &&
-						!notAfter.Add(-minimalCertValidityPeriod).Before(now) {
-						// cert is newer than curBestCert and is not invalid in 7 days
-						longestMatch = nameLength
-						candidate = cert
-					} else if notBefore.Equal(candidate.NotBefore) && !candidate.NotAfter.After(notAfter) {
-						// cert has the same issue date, but is longer valid
-						longestMatch = nameLength
-						candidate = cert
-					} else if notBefore.Before(candidate.NotBefore) &&
-						candidate.NotAfter.Add(-minimalCertValidityPeriod).Before(now) &&
-						notAfter.After(candidate.NotAfter) {
-						// cert is older than curBestCert but curBestCert is invalid in 7 days and cert is longer valid
-						longestMatch = nameLength
-						candidate = cert
-					}
-				case longestMatch > nameLength:
-					if candidate.NotAfter.Add(-minimalCertValidityPeriod).Before(now) &&
-						now.Before(candidate.NotBefore) &&
-						notBefore.Before(now) &&
-						now.Before(notAfter.Add(-minimalCertValidityPeriod)) {
-						// foo.example.org -> *.example.org degradation when NotAfter requires a downgrade
-						longestMatch = nameLength
-						candidate = cert
-					}
-				}
-			}
-		}
-	}
-
-	if longestMatch == -1 {
-		return nil, ErrNoMatchingCertificateFound
-	}
-	return candidate, nil
-}
-
-func prefixGlob(pattern, subj string) bool {
-	// Empty pattern can only match empty subject
-	if pattern == "" {
-		return subj == pattern
-	}
-
-	// If the pattern _is_ a glob, it matches everything
-	if pattern == glob {
-		return true
+func (c *acmClient) newCertDetail(acmDetail *acm.CertificateDetail) *CertDetail {
```

Comment:
Then it can be a specific (non exported) function of the ACM CertificatesProvider